packages/did/src/didAuthorizeExtrinsics/didAuthorizeExtrinsics.ts

```diff
@@ -0,0 +1,145 @@
+/**
+ * Copyright (c) 2018-2022, BOTLabs GmbH.
+ *
+ * This source code is licensed under the BSD 4-Clause "Original" license
+ * found in the LICENSE file in the root directory of this source tree.
+ */
+
+import type { Extrinsic } from '@polkadot/types/interfaces'
+import type { SubmittableExtrinsicFunction } from '@polkadot/api/types'
+
+import { BN } from '@polkadot/util'
+import type {
+  DidVerificationKey,
+  IIdentity,
+  SignCallback,
+  SubmittableExtrinsic,
+  VerificationKeyRelationship,
+} from '@kiltprotocol/types'
+import { SDKErrors } from '@kiltprotocol/utils'
+
+import { FullDidDetails } from '../DidDetails/FullDidDetails.js'
+import {
+  getKeyRelationshipForExtrinsic,
+  increaseNonce,
+} from '../DidDetails/FullDidDetails.utils.js'
+
+import { generateDidAuthenticatedTx } from '../Did.chain.js'
+import { getSigningAlgorithmForVerificationKeyType } from '../Did.utils.js'
+
+/**
+ * Type of a callback used to select one of the key candidates for a DID to sign a given batch of extrinsics.
+ */
+type SelectKeyCallback = (
+  keys: DidVerificationKey[],
+  batch?: Extrinsic[]
+) => Promise<DidVerificationKey>
+
+async function defaultSelectKey([
+  first,
+]: DidVerificationKey[]): ReturnType<SelectKeyCallback> {
+  return first
+}
+
+type GroupedExtrinsics = Array<{
+  extrinsics: Extrinsic[]
+  keyRelationship: VerificationKeyRelationship
+}>
+
+function groupExtrinsicsByKeyRelationship(
+  extrinsics: Extrinsic[]
+): GroupedExtrinsics {
+  const [first, ...rest] = extrinsics.map((extrinsic) => {
+    const keyRelationship = getKeyRelationshipForExtrinsic(extrinsic)
+    if (keyRelationship === 'paymentAccount') {
+      throw new SDKErrors.DidBuilderError(
+        'Can only batch extrinsics that require a DID signature'
+      )
+    }
+    return { extrinsic, keyRelationship }
+  })
+
+  const groups: GroupedExtrinsics = [
+    {
+      extrinsics: [first.extrinsic],
+      keyRelationship: first.keyRelationship,
+    },
+  ]
+
+  rest.forEach(({ extrinsic, keyRelationship }) => {
+    const currentGroup = groups[groups.length - 1]
+    const useCurrentGroup = keyRelationship === currentGroup.keyRelationship
```

Comment:
wouldn't the extrinsics needs to be sorted by keyrelationship for this to work properly?
What if i have `assertionKey -> delegationkey -> assertionKey`?

Reply:
We cannot sort or otherwise change the order of extrinsics, as it affects the result. In the example you have provided we will sign the extrinsics individually without creating batches. Batches will only be created in case there is a sequence of extrinsics all requiring the same key. This is how it was already done in the DidBatchBuilder, and this is the only reason for its existence afaik.

Reply:
ok, let's keep it than to how it is